src/BenchmarksApps/SslStream/Client/Program.cs

```diff
@@ -0,0 +1,236 @@
+    }
+
+    static async Task Run(ClientOptions options)
+    {
+        SetupMeasurements();
+
+        try
+        {
+            switch (options.Scenario)
+            {
+                case Scenario.ReadWrite:
+                    await RunReadWriteScenario(options);
+                    break;
+                case Scenario.Handshake:
+                    await RunHandshakeScenario(options);
+                    break;
+                default:
+                    throw new InvalidOperationException($"Unknown scenario: {options.Scenario}");
+            }
+        }
+        catch (Exception e)
+        {
+            Log($"Exception occured: {e}");
+        }
+    }
+
+    static async Task RunHandshakeScenario(ClientOptions options)
+    {
+        BenchmarksEventSource.Register("sslstream/handshakes/mean", Operations.Avg, Operations.Avg, "Mean TLS handshakes per second.", "Handshakes per second - mean", "n0");
+
+        static async Task EstablishConnection(ClientOptions options)
+        {
+            using var sock = new Socket(SocketType.Stream, ProtocolType.Tcp);
+            await sock.ConnectAsync(options.Hostname, options.Port);
+            using var stream = await EstablishSslStreamAsync(sock, options);
+
+            Interlocked.Increment(ref s_totalHandshakes);
+        }
+
+        s_isRunning = true;
+
+        var task = Task.Run(async () =>
+        {
+            while (s_isRunning)
+            {
+                await EstablishConnection(options);
+            }
+        });
+
+        await Task.Delay(options.Warmup);
+        Log("Completing warmup...");
+
+        Stopwatch sw = Stopwatch.StartNew();
+        Volatile.Write(ref s_totalHandshakes, 0);
+
+        await Task.Delay(options.Duration);
+        s_isRunning = false;
+        Log("Completing scenario...");
+
+        await task;
+        sw.Stop();
+
+        LogMetric("sslstream/handshakes/mean", s_totalHandshakes / sw.Elapsed.TotalSeconds);
+    }
+
+    static async Task RunReadWriteScenario(ClientOptions options)
+    {
+        BenchmarksEventSource.Register("sslstream/read/mean", Operations.Avg, Operations.Avg, "Mean bytes read per second.", "Bytes per second - mean", "n0");
+        BenchmarksEventSource.Register("sslstream/write/mean", Operations.Avg, Operations.Avg, "Mean bytes written per second.", "Bytes per second - mean", "n0");
+
+        static async Task WritingTask(SslStream stream, int bufferSize, CancellationToken cancellationToken)
+        {
+            if (bufferSize == 0)
+            {
+                return;
+            }
+
+            var sendBuffer = new byte[bufferSize];
+            try
+            {
+                while (s_isRunning)
+                {
+                    await stream.WriteAsync(sendBuffer, cancellationToken);
+                    Interlocked.Add(ref s_bytesWritten, bufferSize);
+                }
+            }
+            catch (OperationCanceledException)
```

Comment:
I don't think it's a good idea to use OCE to cancel the last operation 🤔 You are measuring op/s, so I believe it should count the last one as well. Otherwise by not waiting for the last op to finish, you are superficially lowering the result -- because you are adding (part of the) time, but not the bytes. I think it doesn't matter that much that the benchmark itself will run a little bit longer that the asked e.g. 15s, it's better to get more accurate results.